feat: added new connect & exchange pattern first draft

matbmoser opened the pull request:
## Description

This pattern is mentioned here in this issue: #5 

I have added the same content, as a first draft. It can be refactored in the future if needed.

Please add yourself to the authors list if you review the pull request.

The idea is to merge it as a draft so we have already something, and then we can continue to work on it as a community.


## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [X] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [X] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

lgblaumeiser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+## 4. Advantages
+
+- Enables the usage of the EDC in a local machine or private infrastructure.
+  - Allows applications to be deployed in local and that can use the EDC as consumers.
```
what does "use the EDC" means, your point here is, that this allows applications to operate in a private cloud environment or call it as you like, and an connector provisioned in a public cloud environment with only the need to be able to connect from the application to the connector. I am not sure, if this is leading in the right direction, as we face quite a big amount of push scenarios and I am pretty convinced, that the use cases do not like the event mechanism to stop in the connector only because the application is not reachable.

matbmoser replied:
This is a topic that is being worked here:  https://github.com/eclipse-tractusx/sig-release/issues/938

As you probably know. 

I had a talk with Rafael and this pattern will benefit a lot from this functionality, when comes to provider push.

matbmoser replied:
I added a warning, saying this is just valid for the Consumer Pull for now, until the EDR supports provider push.

lgblaumeiser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+
+## 5. Downsides
+
+- Depends on the EDR interface of the EDC.
```
That is an important point, we talk about a feature of the Tractus-X EDC that is not part of Catena-X. At some point, I was thinking of that this proposal is actually a Catena-X proposal, but this opens my eyes here, this only works with the Tractus-X EDC, not even with another upstream derivate of the EDC Connector as the renewal EDR thing is a Tractus-X EDC development. This should be pointed out at some earlier point in the document.

matbmoser replied:
Ok, good point. It is how you use the "Tractus-X Dataspace" not the -X EDC dataspaces in general. 

I believe I had already included this here:
https://github.com/eclipse-tractusx/sig-architecture/pull/6/files#diff-0a3e05bd23eb2c24a4334f7d2a9ac048fa21bdad3363f98999996a465a0e18a5R86

>  In case of Tractus-X we are using the [Tractus-X EDC](https://github.com/eclipse-tractusx/tractusx-edc).

matbmoser replied:
I added the edc url

lgblaumeiser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+- If the policy expires it requires the renegotiation of the contract.
+- If one more policy wants to be accepted the negotiation needs to be repeated, since the conditions have changed. This may cause that some connections are open and not used (but is already less open connections that what we have now)
+- Requires timeout configurations, for waiting for the EDR of the EDC.
+- Would require the development of a new "Simple Data Exchanger" application, maybe a "Kickstart KIT" ??
```
I would say, that at some point providing libraries in multiple languages for using the EDC are the solution, so that an application does not have to redundantly implement all the EDC logic but can make use of a higher level interface.

matbmoser replied:
this is added

lgblaumeiser reviewed (changes requested):
Sorry for having so many comments, I think your idea is a really good one, but we have to work on clarity to get the idea through.

One thought that came to my mind when I read this is, that it is really a good question what kind of document this is. With referring to a non-standardized feature, it is a Tractus-X topic and recommendation for usage of Tractus-X means. But there is a general pattern behind it, which says, that for assets with a character of being used by the same consumer multiple times, long living contracts build a steady connection that should be reused for the length of the contract validity. This is a Catena-X message.

Again this ends up in this void, that we only want to standardize interactions between dataspace participants, but we provide an infrastructure with Tractus-X that is easily misunderstood and actually should not be referred to by use case standards, and only in a very limited way in KITs.

We are currently discussing in CX-Next TAP 7.1 about a general KIT that describes the interaction between two participants and the Industry Core Committe is also planning something along this line. Could be a good embedding for this proposal also towards use cases. 

As it is now, your target audience are the Tractus-X Use Case apps and proprietary app developer that use the Tractus-X EDC as connector implementation.

matbmoser commented:
> As it is now, your target audience are the Tractus-X Use Case apps and proprietary app developer that use the Tractus-X EDC as connector implementation.

I will add this target then

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+
+> You don't need to understand the DSP protocol to use the Tractus-X network as an application
+
+Nowadays, there is a big overhead at the "Application" layer from Tractus-X. Every application is using the Eclipse Dataspace Connector (EDC) in their own way. In case of Tractus-X we are using the [Tractus-X EDC](https://github.com/eclipse-tractusx/tractusx-edc).  The problem is that there is no harmonization in between the consumer and provider applications how they can in the most optimized wa ...
```
An "library" shall be implemented out of this concept (there is one example already but can't be published yet), maybe I am not being clear enough. The idea is that applications do not need to "understand" the DSP protocol, because they should just focus on "do_post" and "do_get" and the rest would be taken care in a layer under the application.

matbmoser replied:
I want to normalize the usage of the DSP.

matbmoser replied:
I have focus better in what you meant.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+
+Therefore, there was built the "EDR" (Endpoint Data Reference) Interface. It was built for simplifying the negotiation and transfer process for the application. Now it is stable since the Jupiter release from Catena-X (TX-EDC >v0.7.5) and is available out there, however there is no usage pattern yet provider for the application side.
+
+Another point is, the EDC still needs to be optimized and several breaking changes are being planned for the future, and therefore all the applications will need to change their way they use the EDC.
```
I wanted to introduce the "EDR" concept in a paragraph, and why it was introduced and can be beneficial now that it is stable. Yes you are right, the motivation shall not come from the short term problems, but it is there for normalizing the adoption of the dataspace, so that this "short term problem" does not becomes a "long term problem".

matbmoser replied:
ok maybe we need to work in the motivation.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+
+Another point is, the EDC still needs to be optimized and several breaking changes are being planned for the future, and therefore all the applications will need to change their way they use the EDC.
+
+For example: If in the future we want to move to the "bring your own identity approach" where everyone brings their DID we need then to expect the applications will care about this.
```
hey this is just an example. in the future will come other things that this pattern will "prevent" the negative impact, which will become a problem..

matbmoser replied:
resolved

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+For example: If in the future we want to move to the "bring your own identity approach" where everyone brings their DID we need then to expect the applications will care about this.
+And that is a huge amount of work for applications to be adapting every release to the latest "base" data space components.
+
+Therefore, this pattern aims to reduce the complexity on the Application layer, easing the adoption of the dataspace for any application using the EDC.
```
I am not sure how to refrase it.

matbmoser replied:
I have refrased in the description.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+
+But in the end you will ALWAYS have applications or services behind the EDC which will respond with information or confirm that the operation was executed.
+
+The only thing that it creates is a "PIPE" in between the companies and the connection remains open.
```
Yes, you understood me! Thats good. You are right is not the motivation, if is placed there then is wrong, this is the "consequence", if you handle your contracts using this pattern you will be able to create "pipes" and remain the "connection" open (reuse the contract and request the token enables a connection) but is not a new connection is the same one, which was "open" when the contract was signed, and the resource was able to be accessed.

matbmoser replied:
I will add two diagrams to explain this better

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+
+Using the EDR interface applications are able to negotiate assets from the catalog without needing to execute a transfer to retrieve the authorization. That was often a problem because the applications always required a "DNS resolvable domain" for the EDR token to be call backed into the application. But with the EDR interface now the application can be deployed in the local machine from the cons ...
+
+Several applications right now are re-doing the negotiation over and over again for every asset that wants to be retrieved, causing the EDC to slow down affecting its performance.
```
ok

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,258 @@
+- A different application wants to be connected (change of asset)
+- The policy expiration time has expired
+
+And once this connection is open I need only the `transferProcessId` to get the authorization from the EDC, because the EDR is already being cached at the EDC, and the token will be kept updated.
```
maybe I should add a chapter break here, to talk about implementation.

matbmoser replied:
Or we could add as "Transfer Process Identification" I think this would it then.

matbmoser replied:
I fixed this

matbmoser commented:
@lgblaumeiser please review it again, I have added some new diagrams that will help to understand the concept.

matbmoser commented on docs/patterns/connect&exchange/media/connect&exchange-context.svg:
In the TRG is still specified we can generate them with argocd, I wanted to do it for more compatibility. But sure I can do if it helps. Love mermaid

matbmoser commented on docs/patterns/connect&exchange/media/connect&exchange-context.svg:
In the context I made the same application, because they need to know the APIs from each other, but, you are right maybe is better to say App B

matbmoser commented on docs/patterns/connect&exchange/media/edc-connector-infoflow.svg:
I can add a simple diagram too
That diagram serves the prurpose of showing the general context on a "mesh" of companies.
But you are right I got the feedback that people additionally want to have it with consumer and provider.

matbmoser replied:
![Architecture-EDC-Service-Library drawio](https://github.com/user-attachments/assets/5fb5ddbc-6d54-4d80-95c4-1eff9b16153d)

matbmoser replied:
I forgot to include this diagram: 
![Architecture-EDC-Service-Library drawio](https://github.com/user-attachments/assets/82a7a662-42c7-442e-957e-87a01c7d7782)

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+
+So the connection keeps open and is really **FAST** (just depends on the content weight and the HTTP overhead).
+
+In the end the level of abstraction is to say, as a consumer, I want to be able to call a "proxy", telling my conditions Policies, Asset to Connect, and then I can make "do_post" or "do_get" and all the connection details will be abstracted.
```
Sure, this could be in the EDC. But maybe is more beneficial to have "library" that does that.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- Reduces the effort for maintenance for apps using the EDC. Just one component needs to be maintained.
+- Easier the adoption of new applications which do not need to include the logic from the EDC.
+- Enables an extra layer of security between the EDC and the data consumer, allowing the consumer to be protected from possible attacks from the EDC.
+- The same "proxy" application can be added in the other side of the EDC and enable an extra "common" security layer to the application.
```
Sorry I meant in the EDC provider side... Behind the provider and the consumer adcs. so that on both sides you have a "layer" that enables the data exchange in a simple way. But I think this is another topic. I will remove this.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- Easier the adoption of new applications which do not need to include the logic from the EDC.
+- Enables an extra layer of security between the EDC and the data consumer, allowing the consumer to be protected from possible attacks from the EDC.
+- The same "proxy" application can be added in the other side of the EDC and enable an extra "common" security layer to the application.
+- Logic can be packed into an EDC extension.
```
And this is the point, every client is implementing their own usage of the EDC. And this is causing performance issues and making the dataspace harder to implement.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- The same "proxy" application can be added in the other side of the EDC and enable an extra "common" security layer to the application.
+- Logic can be packed into an EDC extension.
+- Applications can retrieve data from different applications and shift connections in milliseconds. Example: First I call the DTR, then I call the Submodel Service, then I call the DTR again, then I call the submodel service. All this in milliseconds (which is now not the case)
+- Reduces the vectors of attack to an application behind the EDC.
```
I will just remove it, too hard to argument.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+       - The first EDR negotiation takes 10 seconds and requires an interactive intent to check when the "EDR" is available at the cache, maybe a callback can still be implemented to notify that the data is ready. But in this way a "domain" will be needed from the application side.
+- Requires the usage of an intermediate to talk with the EDC as an application.
+  - Maintenance of this component is required
+  - The maintenance can be complex (requires the EDC developers to maintain it)
```
It was just a suggestion.

matbmoser replied:
ok

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- Requires the usage of an intermediate to talk with the EDC as an application.
+  - Maintenance of this component is required
+  - The maintenance can be complex (requires the EDC developers to maintain it)
+- If the policy expires it requires the renegotiation of the contract.
```
Ok

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+  - Maintenance of this component is required
+  - The maintenance can be complex (requires the EDC developers to maintain it)
+- If the policy expires it requires the renegotiation of the contract.
+- If one more policy wants to be accepted the negotiation needs to be repeated, since the conditions have changed. This may cause that some connections are open and not used (but is already less open connections that what we have now)
```
OK

matbmoser replied:
ok

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- It was supposed that the policies are standardized and are known beforehand, because it was known before the first catalog call.
+- The catalog request could be done using the vanilla "catalog" request or the federated catalog, it is not and will not be specified.
+- This concept just specifies how to handle the "authorization" as a data consumer. The same "application" can be used in both consumer and provider sides, for enabling a harmonized "data consuming" interface.
+- This applies to all the Tractus-X components that use the EDC connector, and enables them to use the EDC in a simple way, allowing the data retrieval to be optimized and harmonized between components. Fostering interoperability.
```
It supports the applications do not need to be updating their compatibility with the EDC every time the edc changes. Maybe, it is fostering "Compatibility". But it could still foster "Interoperability" if it would help applications that are right now not really compatible with the EDC and do not want to implement hard mechanisms that require the understanding of the DSP protocol.

matbmoser replied:
deon

matbmoser commented:
@ndr-brt Thanks for the detailed review!!! 

I will revise your points and I ping you back when I am done :)

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- The Eclipse Dataspace Connector can evolve and be optimized very fast and often without braking changes for the end user application.
+- Reduces the effort for maintenance for apps using the EDC. Just one component needs to be maintained.
+- Easier the adoption of new applications which do not need to include the logic from the EDC.
+- Enables an extra layer of security between the EDC and the data consumer, allowing the consumer to be protected from possible attacks from the EDC.
```
removed.

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- Requires timeout configurations, for waiting for the EDR of the EDC.
+- Would require the development of a new "Simple Data Exchanger" application, maybe a "Kickstart KIT" ??
+  - Or maybe it can be an EDC extension
+- Providing a Library in multiple programming languages would be benefitial to implement this pattern in a harmonized way.
```
done

matbmoser commented:
@ndr-brt I wanted to represent something specifically with the connect and exchange labels in the left, I tried to do it with mermaid but I was not able to do it. Even though I love mermaid I have made it with Draw.io as mentioned in the [TRG 1.4](https://eclipse-tractusx.github.io/docs/release/trg-1/trg-1-4) and it is able to be editable. 


I have included your changes and points, removed the points I could not explain or made no sense, and added the new 1:1 diagram. 

Thanks for the review!

@lgblaumeiser and @ndr-brt could I get an approval? 

In case there is something to change then open another PR :)

matbmoser commented:
I have added you @ndr-brt and indicated the last review date

matbmoser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,268 @@
+- The catalog request could be done using the vanilla "catalog" request or the federated catalog, it is not and will not be specified.
+- This concept just specifies how to handle the "authorization" as a data consumer. The same "application" can be used in both consumer and provider sides, for enabling a harmonized "data consuming" interface.
+- This applies to all the Tractus-X components that use the EDC connector, and enables them to use the EDC in a simple way, allowing the data retrieval to be optimized and harmonized between components. Fostering interoperability.
+- The frequency of the cache "dump" is up to the consumer to decide.
```
I have added upstairs

lgblaumeiser commented on docs/patterns/connect&exchange/README.md:
```diff
@@ -0,0 +1,264 @@
+- A different application wants to be connected (change of asset)
+- The policy expiration time has expired
+
+And once this connection is open I need only the "Transfer Process Identification" to get the authorization from the EDC, because the EDR is already being cached at the EDC, and the token will be kept updated. Then will be just up to the consumer to "dump" the cache when ever it is considered necessary.
```
```suggestion
As soon as the contract negotiation is finalized, the consumer connector based on identifiers related to the contract receive the access token via the EDR interface. Token renewal is done automatically for him, if the available token expires. ```

lgblaumeiser reviewed (approved):
Bascially reformulations to improve readability, hope this is ok for you! Choose as you like.

matbmoser commented:
I will add all of them it's fine for me, thank you for remarking the points!! :)

ndr-brt commented:
@matbmoser my release was still pending, merge was not supposed to happen yet

matbmoser commented:
@ndr-brt https://github.com/eclipse-tractusx/sig-architecture/pull/7

We can still revert it I just need your approval.

ndr-brt commented:
> @ndr-brt #7
> 
> We can still revert it I just need your approval.

no need to revert, it is only a draft document in any case, but for the future it's better follow the good practices and wait for all the pending review to be approved before merging ;)

matbmoser commented:
> > @ndr-brt #7
> > We can still revert it I just need your approval.
> 
> no need to revert, it is only a draft document in any case, but for the future it's better follow the good practices and wait for all the pending review to be approved before merging ;)

Fair enough, sorry for my mistake, please also block the merging next time so I can also know you want to check it before I merge it :)

I will close the revert PR.
